pkg/policy/verifier.go

```diff
@@ -0,0 +1,210 @@
+func Compile(ctx context.Context, v Verification, ww WarningWriter) (Verifier, error) {
+	if err := v.Validate(ctx); err != nil {
+		return nil, err
+	}
+
+	ipc, err := gather(ctx, v, ww)
+	if err != nil {
+		// This should never hit for validated policies.
+		return nil, err
+	}
+
+	return &impl{
+		verification: v,
+		ipc:          ipc,
+		ww:           ww,
+	}, nil
+}
+
+func gather(ctx context.Context, v Verification, ww WarningWriter) (*config.ImagePolicyConfig, error) {
+	pol := *v.Policies
+	ipc := &config.ImagePolicyConfig{
+		Policies: make(map[string]webhookcip.ClusterImagePolicy, len(pol)),
+	}
+
+	for i, p := range pol {
+		var content string
+		switch {
+		case p.Data != "":
+			content = p.Data
+
+		case p.Path != "":
+			raw, err := os.ReadFile(p.Path)
+			if err != nil {
+				return nil, err
+			}
+			content = string(raw)
+
+		case p.URL != "":
+			resp, err := http.Get(p.URL)
+			if err != nil {
+				return nil, err
+			}
+			defer resp.Body.Close()
+			raw, err := io.ReadAll(resp.Body)
+			if err != nil {
+				return nil, err
+			}
+			content = string(raw)
+
+		default:
+			// This should never happen for a validated policy.
+			return nil, fmt.Errorf("unsupported policy shape: %v", p)
+		}
+
+		l, warns, err := ParseClusterImagePolicies(ctx, content)
+		if err != nil {
+			// This path should be unreachable, since we already parse
+			// things during compilation.
+			return nil, fmt.Errorf("parsing policies: %w", err)
+		}
+		if warns != nil {
+			ww("policy %d: %v", i, warns)
+		}
+
+		// TODO(mattmoor): Add additional checks for unsupported things,
+		// like Match, IncludeSpec, etc.
+
+		for _, cip := range l {
+			cip.SetDefaults(ctx)
+			if _, ok := ipc.Policies[cip.Name]; ok {
+				ww("duplicate policy named %q, skipping", cip.Name)
+				continue
+			}
+			// We need to roundtrip the policy through JSON here because
+			// the compiled policy expects to be decoded from JSON and only
+			// sets up certain fields when being unmarshalled from JSON, so
+			// things like keyful verification only work when we roundtrip
+			// through JSON.
+			var compiled webhookcip.ClusterImagePolicy
+			if err := convert(webhookcip.ConvertClusterImagePolicyV1alpha1ToWebhook(cip), &compiled); err != nil {
+				ww("roundtripping policy %v", err)
+				continue
+			}
+			ipc.Policies[cip.Name] = compiled
+		}
+	}
+
+	return ipc, nil
+}
+
+type impl struct {
+	verification Verification
+
+	ipc *config.ImagePolicyConfig
+	ww  WarningWriter
+}
+
+// Check that impl implements Verifier
+var _ Verifier = (*impl)(nil)
+
+// Verify implements Verifier
+func (i *impl) Verify(ctx context.Context, ref name.Reference, kc authn.Keychain) error {
+	tm := getTypeMeta(ctx)
+	om := getObjectMeta(ctx)
+	matches, err := i.ipc.GetMatchingPolicies(ref.Name(), tm.Kind, tm.APIVersion, om.Labels)
+	if err != nil {
+		return err
+	}
+
+	if len(matches) == 0 {
+		switch i.verification.NoMatchPolicy {
+		case "allow":
+			return nil
+		case "warn":
```

Comment:
Generally because most Go code treats normal return values and errors as mutually exclusive, so carrying warnings as wrapped errors when also continuing with normal execution is sort of awkward.  Originally I just logged it, but that presents its own "fun" (I actually need to tweak some of our logging for this to be usable, since as-is some paths spew zap logs 😞 ).

With the `apis.Validatable` pattern, the use of wrapped errors was pragmatic because:
1. It only returns an error, so it doesn't have the "other result" problem,
2. It only deals with `*apis.FieldError`, so we could add methods to mix/separate warnings/errors,
3. This kept us from breaking a several years old API! 😅 

I opted for this because you can trivially accumulate warnings with a little function like:
```
func(s string, i ...interface{}) {
   warns = append(warns, fmt.Sprintf(s, i...))
}
```
... and then we don't have to worry about plumbing anything via return values that virtually every Go programmer will hold wrong. 😅   You can also log it with `log.Printf` or treat them as errors with `log.Fatalf` (I think I use `t.Errorf` in one of the tests 🤓 ).


Happy to discuss more.  Sorry for the wall of text, just trying to dump some of my thinking here as I go through the comments.

Reply:
Makes sense, in other places here, you have a return signature splitting warn/errs, so maybe we could change the return signature to return:
warnings, errors

But, this is not something I'm feeling super strong about, just was thinking from the perspective of the user that wants to easily distinguish if there are errors (already handled), or just warnings. Easy part was not intuitive given the signature of the warning handler function since it only took an interface :)